src/gt4py/next/ffront/decorator.py

```diff
@@ -83,24 +78,40 @@ class Program:
     connectivities: Optional[
         common.OffsetProvider
     ]  # TODO(ricoh): replace with common.OffsetProviderType once the temporary pass doesn't require the runtime information
-
-    # TODO(havogt): pattern will be changed in the compile with variants PR, this pattern is used because a subclass is adding a new field without default
-    _compiled_program: (
-        concurrent.futures.Future[stages.CompiledProgram] | stages.CompiledProgram | None
-    ) = dataclasses.field(init=False, default=None, hash=False, repr=False)
+    enable_jit: bool
+    _static_params: (
+        Sequence[str] | None
+    )  # if the user requests static params, they will be used later to initialize CompiledPrograms
+    _compiled_programs: compiled_program.CompiledPrograms | None = dataclasses.field(
+        default=None, init=False, hash=False, repr=False
+    )
 
     @classmethod
     def from_function(
         cls,
         definition: types.FunctionType,
         backend: next_backend.Backend | None,
         grid_type: common.GridType | None = None,
+        enable_jit: bool = config.ENABLE_JIT,
+        static_params: Sequence[str] | None = None,
         connectivities: Optional[
             common.OffsetProvider
         ] = None,  # TODO(ricoh): replace with common.OffsetProviderType once the temporary pass doesn't require the runtime information
     ) -> Program:
         program_def = ffront_stages.ProgramDefinition(definition=definition, grid_type=grid_type)
-        return cls(definition_stage=program_def, backend=backend, connectivities=connectivities)
+        return cls(
+            definition_stage=program_def,
+            backend=backend,
+            connectivities=connectivities,
+            enable_jit=enable_jit,
+            _static_params=static_params,
+        )
+
+    @property
+    def static_params(self) -> Sequence[str] | None:
+        if self._compiled_programs is not None:
+            return self._compiled_programs.static_params
+        return self._static_params
```

Comment:
If I understand this correctly, `self_static_params` is only used when the static params are provided in the decorator but there is no `self._compiled_programs` yet because no  variant has been yet compiled, right? Wouldn't it be better to always initialize `self._compiled_programs` if there are `static_params`  and then the property should always read from there or return `None`? I think this would also fit well with my other comment about making `self._compiled_programs` a `cached_property`

```suggestion
        if self._compiled_programs is not None:
            return self._compiled_programs.static_params
        return self._static_params
```

Reply:
the current strategy is to initialize `_compiled_program` once all information is available. Currently, this is `static_params` **and** `backend`.
Therefore, we delay instantiation to the point when both is known.